### PR TITLE
Introduce syndication service

### PR DESCRIPTION
**Pull request by aug24**
## What does this change?
Syndication rights are currently received from RCS and stored in ElasticSearch.  This PR provides a primary store for them.

1) Syndication rights and photoshoots are now updated by a dedicated Controller within the metadata service
2) Rights are stored in their own dynamo table; photoshoots continue to be stored with edits.
3) Calculation of affected images takes place within the metadata service, Thrall messages are sent for each affected image
4) Thrall no longer calculates affected images, but _only_ updates ElasticSearch with the provided rights.

A LATER PR will address
1) the issue of Thrall fetching all kinds of data from the primary store (requires more consideration of authentication)
2) removing data from messages

## How can success be measured?
Syndication rights can be reprojected via the syndication store.

## Deployment
See https://trello.com/c/tyarJEax/2241-grid-add-syndication-to-metadata-service

## Tested?
- [ ] locally by committer
- [ ] locally by Guardian reviewer
- [ ] on the Guardian's TEST environment

### CODE REVIEW

File: metadata-editor/app/lib/Syndication.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    case None => Future.successful(Nil)
+    case Some(photoshoot) =>
+      editsStore.scan(config.editsTablePhotoshootIndex, "photoshoot", photoshoot.title)
+        .recover { case NoItemFound => Nil }
+        .map {
+          imageIdsJson => imageIdsJson.map(imageIdJson => imageIdJson.as[String])
+        }
+  }
+
+  def getPhotoshoot(editsStore: EditsStore, id: String)
```
Comment (sihil) [inline metadata-editor/app/lib/Syndication.scala:20]: Minor - perhaps call this `getPhotoshootForImage` as I got a bit confused.

File: thrall/app/lib/kinesis/MessageProcessor.scala [hunk 2]
```diff
@@ -18,25 +19,25 @@ class MessageProcessor(es: ElasticSearch,
...
+      case `updateImageExports` => updateImageExports(updateMessage, logMarker)
+      case `updateImageUserMetadata` => updateImageUserMetadata(updateMessage, logMarker)
+      case `updateImageUsages` => updateImageUsages(updateMessage, logMarker)
+      case `replaceImageLeases` => replaceImageLeases(updateMessage, logMarker)
+      case `addImageLease` => addImageLease(updateMessage, logMarker)
+      case `removeImageLease` => removeImageLease(updateMessage, logMarker)
+      case `setImageCollections` => setImageCollections(updateMessage, logMarker)
+      case `deleteUsages` => deleteAllUsages(updateMessage, logMarker)
+      case `upsertRcsRights` => upsertSyndicationRights(updateMessage, logMarker)
+      case `updateImagePhotoshoot` => updateImagePhotoshoot(updateMessage, logMarker)
```
Comment (sihil) [inline thrall/app/lib/kinesis/MessageProcessor.scala:40]: Lovely.

**Review by sihil — COMMENTED** [review]
A few thoughts. It feels like the business logic of the photoshoots stuff should be testable. If it can be lifted out of the `Future`s there might be a set of functions for calculating usage rights that we can write some tests against. The logic is distributed throughout the controllers at the moment and largely untested as a result of being inside monads.

File: metadata-editor/app/controllers/SyndicationController.scala [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    .foldRight(None: Option[SyndicationRights])((b, a) => (a, b) match {
+      // a exists and has a date, b has a date, compare them
+      case (Some(SyndicationRights(Some(dta), _, _, _)), SyndicationRights(Some(dtb), _, _, _)) => if (dta.isAfter(dtb)) a else Some(b)
+      // a exists, with a date, b doesn't have a date, keep a
+      case (Some(SyndicationRights(Some(_), _, _, _)), _) => a
+      // a is None or doesn't have a date, b has a date, keep b
+      case (_, SyndicationRights(Some(_), _, _, _)) => Some(b)
+      // both are None or have no date, return None.  We never return rights without dates.
+      case _ => None
+    })
```
[the remark this answers is not in the document]
Comment (aug24) [reply metadata-editor/app/controllers/SyndicationController.scala:75]: Gosh!  I spent ages on that fold.  I'll try yours.... when I've written a test.

File: thrall/app/lib/kinesis/MessageProcessor.scala (hunk 2 above)
Comment (aug24) [reply thrall/app/lib/kinesis/MessageProcessor.scala:40]: Thank you :)

File: metadata-editor/app/controllers/EditsController.scala [hunk 4]
```diff
@@ -112,19 +109,30 @@ class EditsController(auth: Authentication, store: EditsStore, notifications: No
...
-    (req.body \ "data").asOpt[Photoshoot].map(photoshoot => {
-      store.jsonAdd(id, "photoshoot", caseClassToMap(photoshoot))
-        .map(publish(id, "update-image-photoshoot"))
-        .map(_ => respond(photoshoot))
+    (req.body \ "data").asOpt[Photoshoot].map(photoshoot => for {
+      js <- store.jsonAdd(id, "photoshoot", DynamoDB.caseClassToMap(photoshoot)) // store - don't care about return
+      imagesInPhotoshoot <- getImagesInPhotoshoot(store, config, Some(photoshoot))     // get list of other images
+    } yield {
+      publish(id, updateImagePhotoshoot)(js)                                           // publish to thrall
+      publish(imagesInPhotoshoot, updateImageSyndication, notifications)               // update all images in photoshoot
```
[the remark this answers is not in the document]
Comment (aug24) [reply metadata-editor/app/controllers/EditsController.scala:117]: Do we?  Only ones which don't have their own rights?

[the remark this answers is not in the document]
Comment (sihil) [reply metadata-editor/app/controllers/EditsController.scala:117]: Only ones without their own rights and only if we've modified the most recent?

We could compute all rights pre-change and post-change and publish the difference?

[the remark this answers is not in the document]
Comment (aug24) [reply metadata-editor/app/controllers/EditsController.scala:117]: Yes.... if we've deleted we could look for more recent rights and not send any updates to the other images if they exist.

[the remark this answers is not in the document]
Comment (sihil) [reply metadata-editor/app/controllers/EditsController.scala:117]: Precisely. Just a thought but thought I'd express the thought here.

**Comment by AWare** [comment]
This looks like it's sending the usual thrall messages when it updates syndication. This might cause quite a lot of extra updates as thrall also figures out which images it wants to update with SyndicationOps. It might be worth having a seperate update message for this.

**Comment by sihil** [comment]
@AWare makes a good point about thrall doing the same thing. What are the next steps in the migration? 

@aug24 It would be really helpful to expand the PR description to include a bit for detail of why we're doing this, why we chose this approach and what the remaining steps are after this PR is merged.

File: metadata-editor/app/lib/Syndication.scala [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  def deleteSyndicationAndPublish(id: String)
+                                 (implicit ec: ExecutionContext): Future[Unit] = {
+    for {
+      // Get the list of ids affected BEFORE doing the delete!
+      photoshootMaybe <- getPhotoshootForImage(id)
+      imagesInPhotoshoot <- getImagesInPhotoshoot(photoshootMaybe)
+      imagesInPhotoshootWithInferredRights <- filterImagesByInference(imagesInPhotoshoot, inferred = true)
+      // Do the delete
+      _ <- syndicationStore.removeKey(id, "photoshoot")
+      _ <- publish(imagesInPhotoshootWithInferredRights, UpdateImageSyndication)
```
[the remark this answers is not in the document]
Comment (sihil) [reply metadata-editor/app/lib/Syndication.scala:42]: Having read further, we're simply calculating the set of IDs that could be modified and the assumption is that Thrall will get the rights for each image when it processes the message??

File: metadata-editor/app/lib/Syndication.scala [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+      .recover { case NoItemFound => JsNull }
+      .map(dynamoEntry =>
+        (dynamoEntry \ syndicationRights).toOption map (x => x.as[SyndicationRights]))
+    )
+  }
+
+  def getMostRecentSyndicationRights(list: List[Option[SyndicationRights]]): Option[SyndicationRights] = list
+    .collect {case Some(sr) => sr}.filter(_.published.nonEmpty).sortBy(_.published.map(-_.getMillis)).headOption
+
+  def filterImagesByInference(ids: List[String], inferred: Boolean)
```
Comment (sihil) [inline metadata-editor/app/lib/Syndication.scala:109]: I'm not quite sure what this is doing. I think we might be caching inferred rights in the syndication store and this is a mechanism for getting stored actual rights or cached inferred rights? If so then I'd challenge the idea that we cache anything as that's a potentially lovely source of bugs. I'd prefer to only store actual rights and then we can do the more expensive calculation as required.

If we want to be performant we could only publish the ID of the directly modified image and then return the complete calculation of all touched images so thrall can do a batch update?

**Review by sihil — COMMENTED** [review]
A few issues. Overall it was quite hard to review this as the PR description lacks any details of the architecture and strategy that was being implemented.

**Review by sihil — COMMENTED** [review]
Happy to do some pairing later (if we can squeeze it in).

File: metadata-editor/app/lib/Syndication.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (aug24) [reply metadata-editor/app/lib/Syndication.scala]: Yes.

File: dev/cloudformation/grid-dev-core.yml [hunk 7]
```diff
@@ -62,19 +62,45 @@ Resources:
...
     Type: AWS::DynamoDB::Table
     Properties:
       TableName: EditsTable
       AttributeDefinitions:
-      - AttributeName: id
-        AttributeType: S
+        - AttributeName: id
+          AttributeType: S
+        - AttributeName: photoshoot
+          AttributeType: S
```
[the remark this answers is not in the document]
Comment (aug24) [reply dev/cloudformation/grid-dev-core.yml:87]: There's no mechanism to mark it optional, and it already works in CODE and PROD :)

**Comment by aug24** [comment]
> This looks like it's sending the usual thrall messages when it updates syndication. 

Now that we have a plan for delivery, those lines are temporarily removed.  We will re-introduce them as per the deployment plan (see https://trello.com/c/tyarJEax/2241-grid-add-syndication-to-metadata-service)

**Review by jonathonherbert — APPROVED** [review]
Code looks good, comments look addressed – in the absence of reviewers revisiting this code , given our deployment plan it looks like there's plenty of room to fix forward. Testing discussed out of band, logic looks tested, I'm a fan of integration tests for this sort of service (via localstack or similar) but it's not a blocker.

**Pull request merged**